File: source/lib/theme_healy.py
"""
theme_healy.py  --  Matplotlib / Seaborn defaults following Kieran Healy's principles

Based on:
    Healy, Kieran. 2018. *Data Visualization: A Practical Introduction.*
    Princeton University Press. https://socviz.co/

Key principles applied here (see Healy ch. 1, 3, 8):
    - Maximize data-to-ink ratio: remove top/right spines, heavy borders,
      and background fills (Healy 2018, ch. 8)
    - Use ColorBrewer "Set2" for categorical data -- perceptually uniform
      and colorblind-safe (ch. 1, 8)
    - Use viridis for continuous scales -- perceptually ordered and
      accessible across color-vision deficiencies (ch. 1)
    - Remove minor gridlines; keep major gridlines light (ch. 8)
    - Favor faceting (small multiples) over cramming groups into one
      plot with too many colors (ch. 4)
    - Always label axes clearly; include titles and source captions (ch. 1, 3)
    - Export at publication quality with explicit dimensions (ch. 8)

Import once at the top of any analysis notebook or script:
    from source.lib.theme_healy import apply_theme, PALETTE, healy_save

What this does:
    1. apply_theme() sets matplotlib rcParams for a clean, minimal look
    2. PALETTE provides a colorblind-safe categorical palette (ColorBrewer Set2)
    3. healy_save() saves figures with sensible defaults
"""

import matplotlib as mpl
import matplotlib.pyplot as plt

# ---------------------------------------------------------------------------
# 1. Categorical palette -- ColorBrewer "Set2" (colorblind-safe, up to 8)
# ---------------------------------------------------------------------------
PALETTE = [
    "#66c2a5", "#fc8d62", "#8da0cb", "#e78ac3",
    "#a6d854", "#ffd92f", "#e5c494", "#b3b3b3",
]

# ---------------------------------------------------------------------------
# 2. Apply theme (call once at top of script / notebook)
# ---------------------------------------------------------------------------
def apply_theme(font_size=12):
    """Set matplotlib rcParams following Healy's principles."""
    mpl.rcParams.update({
        # Figure
        "figure.figsize":       (6, 4),
        "figure.dpi":           150,
        "figure.facecolor":     "white",
        "savefig.dpi":          300,
        "savefig.bbox":         "tight",
        "savefig.facecolor":    "white",

        # Font
        "font.size":            font_size,
        "font.family":          "sans-serif",
        "axes.titlesize":       font_size * 1.15,
        "axes.titleweight":     "bold",
        "axes.labelsize":       font_size * 0.9,

        # Axes
        "axes.spines.top":      False,
        "axes.spines.right":    False,
        "axes.edgecolor":       "grey70",
        "axes.linewidth":       0.6,
        "axes.grid":            True,
        "axes.grid.which":      "major",
        "axes.prop_cycle":      mpl.cycler(color=PALETTE),

        # Grid
        "grid.color":           "#e5e5e5",
        "grid.linewidth":       0.4,

        # Ticks
        "xtick.color":          "grey30",
        "ytick.color":          "grey30",
        "xtick.labelsize":      font_size * 0.8,
        "ytick.labelsize":      font_size * 0.8,

        # Legend
        "legend.frameon":       False,
        "legend.fontsize":      font_size * 0.8,
    })

    # If seaborn is available, set its defaults too
    try:
        import seaborn as sns
        sns.set_palette(PALETTE)
    except ImportError:
        pass


# ---------------------------------------------------------------------------
# 3. Save helper
# ---------------------------------------------------------------------------
def healy_save(fig, filepath, width=6, height=4, dpi=300, **kwargs):
    """Save a figure with publication-quality defaults."""
    fig.set_size_inches(width, height)
    fig.savefig(filepath, dpi=dpi, bbox_inches="tight",
                facecolor="white", **kwargs)

File: source/lib/theme_healy.R
# ===========================================================================
# theme_healy.R  --  ggplot2 defaults following Kieran Healy's principles
#
# Based on:
#   Healy, Kieran. 2018. *Data Visualization: A Practical Introduction.*
#   Princeton University Press. https://socviz.co/
#
# Key principles applied here (see Healy ch. 1, 3, 8):
#   - Use theme_minimal() as a base; avoid heavy borders and backgrounds
#     (Healy 2018, ch. 8: "less non-data ink is almost always better")
#   - Prefer ColorBrewer palettes for categorical data; they are designed
#     to be perceptually uniform and colorblind-safe (ch. 1, 8)
#   - Use viridis for continuous scales -- perceptually ordered and
#     accessible across color-vision deficiencies (ch. 1)
#   - Remove minor gridlines; keep major gridlines light (ch. 8)
#   - Favor faceting (small multiples) over overloading a single plot
#     with too many colors or groups (ch. 4)
#   - Always label axes clearly and include a source caption (ch. 1)
#   - Use labs() for titles/subtitles/captions; let the data speak (ch. 3)
#   - Export at publication quality with explicit dimensions (ch. 8)
#
# Source once at the top of any analysis script:
#   source(here::here("source/lib/theme_healy.R"))
#
# What this does:
#   1. Defines theme_healy() built on theme_minimal()
#   2. Sets it as the session default via theme_set()
#   3. Provides palette helpers (scale_color_healy / scale_fill_healy)
#   4. Provides healy_save() -- a ggsave wrapper with sensible defaults
# ===========================================================================

library(ggplot2)
library(scales)

# ---------------------------------------------------------------------------
# 1. Custom theme
# ---------------------------------------------------------------------------
theme_healy <- function(base_size = 12, base_family = "") {
  theme_minimal(base_size = base_size, base_family = base_family) %+replace%
    theme(
      # Clean plot area
      panel.grid.minor   = element_blank(),
      panel.grid.major   = element_line(color = "grey90", linewidth = 0.3),

      # Axes
      axis.title         = element_text(size = rel(0.9), face = "plain"),
      axis.text          = element_text(size = rel(0.8), color = "grey30"),
      axis.ticks         = element_line(color = "grey70", linewidth = 0.3),

      # Legend: top by default, no box
      legend.position    = "top",
      legend.title       = element_text(size = rel(0.85)),
      legend.text        = element_text(size = rel(0.8)),
      legend.key         = element_blank(),

      # Titles
      plot.title         = element_text(size = rel(1.15), face = "bold",
                                        hjust = 0, margin = margin(b = 6)),
      plot.subtitle      = element_text(size = rel(0.9), color = "grey40",
                                        hjust = 0, margin = margin(b = 10)),
      plot.caption       = element_text(size = rel(0.7), color = "grey50",
                                        hjust = 1, margin = margin(t = 8)),
      plot.title.position = "plot",

      # Strip (facet) labels
      strip.text         = element_text(size = rel(0.85), face = "bold"),
      strip.background   = element_blank()
    )
}

# Set as session default
theme_set(theme_healy())

# ---------------------------------------------------------------------------
# 2. Color palettes  (ColorBrewer / viridis -- colorblind-safe)
# ---------------------------------------------------------------------------

# Discrete: qualitative ColorBrewer "Set2" (up to 8 categories)
scale_color_healy <- function(...) {
  scale_color_brewer(palette = "Set2", ...)
}
scale_fill_healy <- function(...) {
  scale_fill_brewer(palette = "Set2", ...)
}

# Continuous: viridis (perceptually uniform, colorblind-safe)
scale_color_healy_c <- function(...) {
  scale_color_viridis_c(option = "viridis", ...)
}
scale_fill_healy_c <- function(...) {
  scale_fill_viridis_c(option = "viridis", ...)
}

# ---------------------------------------------------------------------------
# 3. ggsave wrapper with publication defaults
# ---------------------------------------------------------------------------
healy_save <- function(filename, plot = last_plot(),
                       width = 6, height = 4, dpi = 300, ...) {
  ggsave(filename, plot = plot, width = width, height = height,
         dpi = dpi, bg = "white", ...)
}
